chrome/content/zotero/xpcom/itemTreeManager.js

```diff
@@ -0,0 +1,222 @@
+/*
+    ***** BEGIN LICENSE BLOCK *****
+    
+    Copyright © 2019 Corporation for Digital Scholarship
+                     Vienna, Virginia, USA
+                     https://digitalscholar.org
+    
+    This file is part of Zotero.
+    
+    Zotero is free software: you can redistribute it and/or modify
+    it under the terms of the GNU Affero General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+    
+    Zotero is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Affero General Public License for more details.
+    
+    You should have received a copy of the GNU Affero General Public License
+    along with Zotero.  If not, see <http://www.gnu.org/licenses/>.
+    
+    ***** END LICENSE BLOCK *****
+*/
+
+const { DEFAULT_COLUMNS } = require("zotero/itemTreeColumns");
+
+/**
+ * @typedef {import("../itemTreeColumns.jsx").ItemTreeColumnOption} ItemTreeColumnOption
+ */
+
+class ItemTreeManager {
+    _observerAdded = false;
+    /** @type {ItemTreeColumnOption[]} */
+    _customColumns = [];
+    constructor() {
+    }
+
+    /** 
+     * Register a custom column. All registered columns must be valid, and must have a unique dataKey.
+     * @param {ItemTreeColumnOption | ItemTreeColumnOption[]} option
+     * @returns {boolean} Although it's async, resolving does not promise the item trees are updated.
+     * @throws {Error} If the column option is missing required properties
+     * @throws {Error} If the column is already registered
+     */
+    async registerColumn(option) {
+        const success = this._addColumn(option);
+        if (!success) {
+            return false;
+        }
+        this._ensureObserverAdded();
+        await this._reset();
+        return true;
+    }
+
+    /**
+     * Unregister a custom column
+     * @param {string} dataKey - The dataKey of the column to unregister
+     * @returns {boolean} Although it's async, resolving does not promise the item trees are updated.
+     */
+    async unregisterColumn(dataKey) {
+        const success = this._removeColumn(dataKey);
+        if (!success) {
+            return false;
+        }
+        await this._reset();
+        return true;
+    }
+
+    // TODO: add cell renderer registration
+
+    /**
+     * Get all registered columns
+     */
+    get columns() {
+        return [...DEFAULT_COLUMNS, ...this._customColumns];
+    }
```

Comment:
Seems like this should be either a method on an instance of a specific itemTree or a function that takes the item tree id as a parameter, and gets the columns used in that specific item tree. But given that there's little use for plugins for this function and it's primarily to be used by itemTree code to figure out which columns it should display, it should probably be the latter and only return the custom columns. So something like `getCustomColumns(itemTreeId)`.